### ISSUE TITLE
[QUESTION] force_send_socket socket selection

### ISSUE THREAD

**Issue by attermann** [OpenSIPS/opensips#2083]
**OpenSIPS version you are running**
```
version: opensips 2.4.7 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 495b980
main.c compiled on 22:17:41 Apr  7 2020 with gcc 5.4.0
Running on Ubuntu 16.04
```

Not certain if this is a bug per-se, a limitation, or something else entirely.  My proxy/registrar is configured to listen on multiple ports of the same interface, and I use these ports in my config to detect whether a request or response originates internally or externally, and whether or not transcoding should be configured.

To be specific, I use port 5060 to communicate with external UAs and 6050 to communicate with internal UAs *without* transocding, and port 7080 to communicate with external UAs and 8070 to communicate with internal UAs *with* transocding.  I am using force_send_socket() to choose which port a request or response goes out on.  Most traffic is TCP.

Recently I have discovered that sometimes when a particular server registers clients with opensips, two separate applications running on that server (both registering to opensips) can use the same TCP port when they are connecting to different ports on opensips (since these are client connections which can re-use a port as long the destination ip:port is different).

So for example, assuming my opensips is listening on tcp:1.1.1.1:5060 and tcp:1.1.1.1:7080, the server might register client A from tcp:2.2.2.2:1234 to opensips at tcp:1.1.1.1:5060, and client B from tcp:2.2.2.2:1234 (same source protocol:ip:port) to opensips at tcp:1.1.1.1:7080 (same destination protocol:ip but different port).

This is all fine and good so far, but when it comes time to send a call to client B my script would use force_send_socket(tcp:1.1.1.1:7080) to send from port 7080 to the server at tcp:2.2.2.2:1234, but the application that registered client A would actually receive the call instead.  I don't know if force_send_socket() is not selecting the correct socket or if the issue is something different but I wanted to post my question here in case anybody that is familiar with the internal workings of force_send_socket() may know what's happening here.

Thanks in advance,

Chad.

**Comment by attermann** [comment]
Also note, I chose not to pass socket_info to tcp_conn_get() in modules/tls_mgm/tls_select.c:get_cur_connection() because I was was unsure about potential unintended side-effects.

```tcp_conn_get(msg->rcv.proto_reserved1, 0, 0, PROTO_NONE, &c, NULL/*fd*/, NULL/*send_sock*/);```

TODO: Evaluate if msg force_send_socket or another socket_info should be passed from here.

**Comment by bogdan-iancu** [comment]
Need to allocate time and review the scenario described here.

**Comment by bogdan-iancu** [comment]
@attermann, I guess you figured out the response by yourself, and yes your conclusion is correct. The TCP conn selection depends only on the destination IP+port and it is totally ignoring the source/local socket.

This is to have a final answer here.

For the fix, let move on #2090
